core/src/main/java/io/parsingdata/metal/data/ConstantSource.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.data;
+
+import static io.parsingdata.metal.Util.bytesToHexString;
+import static io.parsingdata.metal.Util.checkNotNull;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+public class ConstantSource extends Source {
+
+    private final byte[] data; // Private because array content is mutable.
+
+    public ConstantSource(byte[] data) {
```

Comment:
Quite a few missing `final` statements throughout this new code (also in all new `equals` methods).

*Hint*: you can configure the IntelliJ "inspections" to mark missing `final`s as **error**. That way you will never miss one again.